api/src/main/java/org/xwiki/apps/googleapps/internal/GoogleAppsManagerImpl.java

```diff
@@ -0,0 +1,1372 @@
+        } catch (IOException e) {
+            String msg = "Can't read build time.";
+            log.warn(msg, e);
+            throw new RuntimeException(msg, e);
+        }
+    }
+
+
+    // from ActiveDirectorySetupListener
+
+    /**
+     * Note that this dispose() will get called when this Extension is uninstalled which is the use case we want to
+     * serve. The fact that it'll also be called when XWiki stops is a side effect that is ok.
+     */
+    @Override
+    public void dispose()
+    {
+        XWiki xwiki = getXWiki();
+        // XWiki can be null in the case when XWiki has been started and not accessed (no first request done and thus
+        // no XWiki object initialized) and then stopped.
+        if (xwiki != null) {
+            // Unset the Authentication Service (next time XWiki.getAuthService() is called it'll be re-initialized)
+            xwiki.setAuthService(null);
+        }
+    }
+
+
+    // -----------------------------------------------------------------------------------------
+    private XWiki getXWiki()
+    {
+        XWiki result = null;
+        XWikiContext xc = this.xwikiContextProvider.get();
+        // XWikiContext could be null at startup when the Context Provider has not been initialized yet (it's
+        // initialized after the first request).
+        if (xc != null) {
+            result = xc.getWiki();
+        }
+        return result;
+    }
+
+    // ------------------------- public API ---------------------------------------------
+
+    /**
+     *
+     * @return if the app is configured to use the Google Drive integration (default: yes).
+     * @since 3.0
+     */
+    @Unstable
+    public boolean isUseDrive() {
+        return configScopeUseDrive;
+    }
+
+
+
+
+    // ----------------------------- Google Apps Tool (mostly request specific) -----------------------------------
+
+    private  String getOAuthUrl() throws XWikiException {
+        XWikiContext context = xwikiContextProvider.get();
+        DocumentReference oauthReference = new DocumentReference(context.getWikiId(),
+                SPACENAME, OAUTH);
+        return getXWiki().getDocument(oauthReference, context).getExternalURL(VIEWACTION, context);
+    }
+
+    private DocumentReference getXWikiUserClassRef() {
+        return new DocumentReference(WIKINAME, XWIKISPACE, "XWikiUsers");
+    }
+
+    private String getCurrentXWikiUserName() {
+        DocumentReference userDoc = xwikiContextProvider.get().getUserReference();
+        String uName = userDoc == null ? XWIKIGUEST : userDoc.getName();
+        if (XWIKIGUEST.equals(uName)) {
+            uName = uName + "-" + xwikiContextProvider.get().getRequest().getSession().hashCode();
+        }
+        return uName;
+    }
+
+    /**
+     * @param userName the name of the user
+     * @return A DocumentReference for the given username.
+     * @since 3.0
+     */
+    @Unstable
+    public DocumentReference createUserReference(String userName) {
```

Comment:
You need to use the user resolver.

Reply:
But this is done inside and includes a bit more cleanup logic.
Are you telling me I should use a "user-resolver" or something different?
paul

Reply:
Yes, you will end up using something like `userResolver.resolve(username)` and that will replace the entire method. No more `if-statements` and no more string operations.

Reply:
Done. Thanks.